### PR TITLE
test: add test-containerd 

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**- What I did**

Added test for containerd, especially for snapshotters.
cc @dmcgowan https://github.com/containerd/containerd/issues/767

**- How I did it**

Added `test/pkg/containerd/Dockerfile`

**- How to verify it**

```console
$ docker build -t test-containerd test/pkg/containerd
$ cd test/cases/040_packages/002_containerd
$ moby build test-containerd
$ linuxkit run hyperkit -mem 2048 test-containerd
```

Note:
- requires a fix for `moby build` https://github.com/moby/tool/pull/44
- kernel quietly mount broken rootfs when `-mem` is too small (Is this expected?)
- I didn't put the image digest to the YAML, but I'll update the YAML when the image is pushed to the hub

**- Description for the changelog**

test: add test-containerd

**- A picture of a cute animal (not mandatory but encouraged)**
penguin

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

### CODE REVIEW

File: test/pkg/containerd/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# alpine 3.5 lacks go1.8
+FROM golang:1.8
```
Comment (rneugeba) [inline test/pkg/containerd/Dockerfile]: If you use the `linuxkit/alpine` package you get golang 1.8. See for example: https://github.com/linuxkit/linuxkit/blob/master/pkg/sysctl/Dockerfile

**Comment by rneugeba** [comment]
I'm not sure we want to add containerd tests to LinuxKit. Ideally, containerd would be tested separately...and ideally with LinuxKit :)

**Comment by justincormack** [comment]
"kernel quietly mount broken rootfs when -mem is too small (Is this expected?)" - yes that is a known issue at present...

File: test/pkg/containerd/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+RUN apt update && apt install -y btrfs-tools
+ENV CONTAINERD_COMMIT=199544ea8009bf6988a81456c7dec66b2d3268e0
```
Comment (justincormack) [inline test/pkg/containerd/Dockerfile]: I was thinking of moving the `containerd` git checkout into the `linuxkit/alpine` package then we could use the same one.

**Comment by justincormack** [comment]
@rneugeba we should test the `containerd` we have installed. But I would rather we specifically ran the test suite against the binary installed in the system.

**Comment by rneugeba** [comment]
I agree

**Comment by justincormack** [comment]
@AkihiroSuda it would be nice to add this once the issues are addressed...

**Comment by AkihiroSuda** [comment]
I'll try to update PR in a few days.. Sorry for being late 🙏

**Comment by AkihiroSuda** [comment]
Unfortunately race detector does not seem to work on alpine.. (requires glibc)
https://github.com/golang/go/issues/14481
https://github.com/linuxkit/linuxkit/compare/master...AkihiroSuda:containerd-test.alpine.broken?expand=1

**Comment by AkihiroSuda** [comment]
@justincormack 

Which do you prefer:
1.  this PR as-is (debian-based)
2. update upstream containerd Makefile to disable race detector, so that it won't require glibc

**Comment by justincormack** [comment]
Hmm, I think we need to test the installed `containerd` not build a new one in LinuxKit, so I don't see that we can run the race tests anyway as they require a rebuild.

**Comment by AkihiroSuda** [comment]
Ideally yes, but snapshotter tests are implemented as daemon-less UT, and rebuilding for Linuxkit seems inevitable.

So I'd like to make this forward with the plan 2.

**Comment by AkihiroSuda** [comment]
Updated PR

- Based on alpine
- `CONTAINERD_COMMIT` is set in `tools/alpine/Makefile`
- Race detector is disabled so as to support alpine
- Test failure will be fixed when https://github.com/containerd/containerd/pull/996 gets merged

```
panic: test executed panic(nil) or runtime.Goexit

goroutine 6 [running]:
testing.tRunner.func1(0xc4200575f0)
	/usr/lib/go/src/testing/testing.go:622 +0x29f
runtime.Goexit()
	/usr/lib/go/src/runtime/panic.go:366 +0x157
testing.(*common).FailNow(0xc420057450)
	/usr/lib/go/src/testing/testing.go:486 +0x3b
testing.(*common).Fatal(0xc420057450, 0xc420036c50, 0x1, 0x1)
	/usr/lib/go/src/testing/testing.go:524 +0x71
github.com/containerd/containerd/snapshot/btrfs.setupBtrfsLoopbackDevice(0xc420057450, 0xc420132c60, 0x28, 0xc420132c60)
	/root/go/src/github.com/containerd/containerd/snapshot/btrfs/btrfs_test.go:162 +0x410
github.com/containerd/containerd/snapshot/btrfs.boltSnapshotter.func1(0x560383db46e0, 0xc420132ba0, 0xc420132c60, 0x28, 0xc420132ba0, 0x23, 0x0, 0x0, 0x15)
	/root/go/src/github.com/containerd/containerd/snapshot/btrfs/btrfs_test.go:27 +0x4f
...
```

**Comment by justincormack** [comment]
Btrfs should work but it needs to be modprobe first at present until we fix
module autoload.

**Comment by AkihiroSuda** [comment]
Updated again (added `util-linux` for `losetup` and `btrfs-progs` for `mkfs.btrfs`).

Now test passes.

**Comment by rn** [comment]
the `btrfs` kernel module does seem to get autoloaded, see: https://github.com/linuxkit/linuxkit/issues/1742#issuecomment-308257147

so there may be subtle difference in setup...

File: test/pkg/containerd/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  cd $GOPATH/src/github.com/containerd && \
+  git clone https://github.com/containerd/containerd.git
+WORKDIR $GOPATH/src/github.com/containerd/containerd
+# CONTAINERD_COMMIT is defined in linuxkit/alpine
+RUN git checkout $CONTAINERD_COMMIT
+
+# unset -race (does not work on alpine; see golang/go#14481)
+ENV TESTFLAGS=
+ENTRYPOINT ["make", "root-test"]
+LABEL org.mobyproject.config='{"net": "host", "capabilities": ["all"], "tmpfs": ["/tmp:exec"], "binds": ["/dev:/dev"]}'
```
Comment (rn) [inline test/pkg/containerd/Dockerfile]: maybe you need to add `"/sys:/sys"` to the bind mounts for the `btrfs` autoload to work?

**Comment by AkihiroSuda** [comment]
Seems working as-is
```
 - 000-sysctl
🐳 root-test
[    5.554243] clocksource: Switched to clocksource tsc
?   	github.com/containerd/containerd/snapshot	[no test files]
[   14.277977] raid6: sse2x1   gen() 10863 MB/s
[   14.446358] raid6: sse2x1   xor()  7692 MB/s
[   14.618086] raid6: sse2x2   gen() 14393 MB/s
[   14.788199] raid6: sse2x2   xor()  9469 MB/s
[   14.955850] raid6: sse2x4   gen() 14707 MB/s
[   15.127829] raid6: sse2x4   xor() 10805 MB/s
[   15.297585] raid6: avx2x1   gen() 21495 MB/s
[   15.467423] raid6: avx2x2   gen() 27051 MB/s
[   15.637203] raid6: avx2x4   gen() 25318 MB/s
[   15.637528] raid6: using algorithm avx2x2 gen() 27051 MB/s
[   15.637899] raid6: using avx2x2 recovery algorithm
[   15.641599] xor: automatically using best checksumming function   avx
[   15.651644] Btrfs loaded, crc32c=crc32c-intel
[   15.652151] BTRFS: device fsid 449d5ca9-9ffe-4834-91de-8d20f34d920e devid 1 transid 5 /dev/loop0
[   15.665627] BTRFS info (device loop0): disk space caching is enabled
[   15.666017] BTRFS info (device loop0): has skinny extents
[   15.666308] BTRFS info (device loop0): flagging fs with big metadata feature
[   15.667363] BTRFS info (device loop0): creating UUID tree
[   16.104534] BTRFS: device fsid 48523579-12a5-4e31-90cd-bd0ff648e4e5 devid 1 transid 5 /dev/loop0
[   16.185316] BTRFS info (device loop0): disk space caching is enabled
[   16.185863] BTRFS info (device loop0): has skinny extents
[   16.186289] BTRFS info (device loop0): flagging fs with big metadata feature
[   16.187832] BTRFS info (device loop0): creating UUID tree
[   16.562612] BTRFS: device fsid 88d6702e-dee4-4249-a087-66cf52c8a3ab devid 1 transid 5 /dev/loop0
[   16.645583] BTRFS info (device loop0): disk space caching is enabled
[   16.646946] BTRFS info (device loop0): has skinny extents
[   16.648125] BTRFS info (device loop0): flagging fs with big metadata feature
[   16.651088] BTRFS info (device loop0): creating UUID tree
[   17.021603] BTRFS: device fsid 86aaf00e-556e-4f8b-afc1-1fc7a1655780 devid 1 transid 5 /dev/loop0
[   17.105198] BTRFS info (device loop0): disk space caching is enabled
[   17.106203] BTRFS info (device loop0): has skinny extents
[   17.107355] BTRFS info (device loop0): flagging fs with big metadata feature
[   17.109934] BTRFS info (device loop0): creating UUID tree
[   17.499642] BTRFS: device fsid 8cd88fde-5466-4ebd-a99b-73192a6e6661 devid 1 transid 5 /dev/loop0
[   17.586794] BTRFS info (device loop0): disk space caching is enabled
[   17.588012] BTRFS info (device loop0): has skinny extents
[   17.589121] BTRFS info (device loop0): flagging fs with big metadata feature
[   17.591798] BTRFS info (device loop0): creating UUID tree
[   17.958572] BTRFS: device fsid b74ed4b6-8fb8-4fcd-a699-bf866779eb9d devid 1 transid 5 /dev/loop0
[   18.045559] BTRFS info (device loop0): disk space caching is enabled
[   18.047069] BTRFS info (device loop0): has skinny extents
[   18.048158] BTRFS info (device loop0): flagging fs with big metadata feature
[   18.051225] BTRFS info (device loop0): creating UUID tree
ok  	github.com/containerd/containerd/snapshot/btrfs	4.422s
ok  	github.com/containerd/containerd/snapshot/naive	0.013s
ok  	github.com/containerd/containerd/snapshot/overlay	0.016s
ok  	github.com/containerd/containerd/snapshot/storage	0.008s
?   	github.com/containerd/containerd/snapshot/storage/proto	[no test files]
?   	github.com/containerd/containerd/snapshot/testsuite	[no test files]
 - 001-test
[   51.384305] ACPI: Preparing to enter system sleep state S5
[   51.385328] reboot: Power down
```

**Comment by AkihiroSuda** [comment]
CI failure is just because I didn't (couldn't) push and determine digests of images

File: tools/alpine/Dockerfile [hunk 4]
```diff
@@ -54,3 +54,5 @@ COPY --from=shellcheck /usr/local/bin/shellcheck /usr/local/bin/shellcheck
 COPY --from=shellcheck /usr/local/lib/ /usr/local/lib/
 
 RUN apk update && apk upgrade -a
+
+ARG CONTAINERD_COMMIT=bdf9f5f7388e8203e63a74b89800f7f3dd4a7743
```
Comment (rn) [inline tools/alpine/Dockerfile]: @justincormack what do you think of adding the commit SHA to alpine base. I kinda like it. and we could also add:
```
ENV DOCKER_BUCKET get.docker.com
ENV DOCKER_VERSION 17.05.0-ce
ENV DOCKER_SHA256 340e0b5a009ba70e1b644136b94d13824db0aeb52e09071410f35a95d94316d9
```
to it as we likely need this in a number of different packages

Alternatively we could add the source code and/or binaries to the alpine base

**Comment by rn** [comment]
@AkihiroSuda could you rebase? This LGTM afterwards

**Comment by AkihiroSuda** [comment]
rebased

**Comment by rn** [comment]
closing in favour of #2062

**Pull request closed without merging**